V2.0

kriszyp opened the pull request.

KyleAMathews commented:
Don't want to spoil the surprise necessarily but curious what the theme of 2.0 is!?

kriszyp commented:
I've been meaning to start documenting what will be new/change, so thanks for asking (and I will probably translate these to release notes once it is ready):

The most central change is a rewrite of how data is batched and written to LMDB. In v1, all writes are accumulated in an array and an event is queued to submit that array to the worker thread, at which point all the keys are serialized into a structure that the writer thread can consume. In v2.0, all writes are immediately serialized to a binary format that the writer thread can consume. This have multiple implications (and hopefully benefits):

- The writer thread can begin writing pending puts/deletes to LMDB (and async-txns can be queued) _immediately_ after they have been called (in parallel to the main thread), rather having to wait for a future event to submit the writes and start the transaction. And as more writes occur, they can continue to be queued and consumed by the writer thread as long as it is still processing a transaction. This should provide better write latency and flexibility in batching.
- This should benefit memory management, allowing pending writes to written and be cleared/collected from memory without having to wait for transactions to complete and events to be queued, which should be more robust for heavy writing situations, preventing overloading of queued operations.
- All keys are immediately serialized using JavaScript. The (async) put/delete operations don't need to depend on a C++ for writing keys, so any errors in serializing keys are immediately triggered.
- Because all key (de)-serialization happens in JS, lmdb-store can now support custom (user-provided) key encodings (so for example, if you want a custom UUID encoder that can efficiently encode UUIDs in 16 bytes).
- Some additional put flags can now be supported (like "append" puts) (#43).
- The V8 "fast api" calls can be used for synchronous puts/deletes (used within async-txns).
- Added some functions for additional explicit batching.

The codebase has been upgraded to ESM (and refactored to be more modular), so lmdb-store can be used with native ESM loading (of course, built CJS modules are included too). This is also one of the necessary steps in hopefully supporting Deno (#83) at some point (once their foreign function interface is complete enough to be useable).

A new binary key comparison function is used that can do (32-bit) word by word comparison, instead of byte by byte comparison, for a little faster b-tree traversals.

Some other ongoing work, that may or may not be included is experimenting with a modification to LMDB to support "overlapping" syncs, and a compatibility layer for upgrading from LevelUp.

Changes to be aware of:
There aren't any intentional breaking changes to the (documented) API itself. Probably, the most significant potential "breaking" change is that by default, writes will not be guaranteed to be in the transaction just because they are in the same event turn. Writes are included in the same transaction if they explicitly batched together (using batch, ifVersion, or transactionAsync), or the event turn batching behavior can be enabled as an option. There will also be some cleanup of deprecated/undocumented functions.

This is probably about 60% done. Most basic stuff works, but there is plenty of logging still in place as I work through some of the more complicated stuff.
